staging/src/k8s.io/client-go/rest/exec.go

```diff
@@ -0,0 +1,58 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package rest
+
+import (
+	"fmt"
+
+	"k8s.io/client-go/pkg/apis/clientauthentication"
+	clientauthenticationapi "k8s.io/client-go/pkg/apis/clientauthentication"
+)
+
+// This file contains Config logic related to exec credential plugins.
+
+// ConfigToExecCluster creates a clientauthenticationapi.Cluster with the corresponding fields from
+// the provided Config. If config.ExecProvider is nil or config.ExecProvider.ProvideClusterInfo is
+// false, then this function will return an empty clientauthenticationapi.Cluster.
+func ConfigToExecCluster(config *Config) (*clientauthenticationapi.Cluster, error) {
+	if config.ExecProvider == nil || !config.ExecProvider.ProvideClusterInfo {
+		return &clientauthentication.Cluster{}, nil
+	}
+
+	caData, err := dataFromSliceOrFile(config.CAData, config.CAFile)
+	if err != nil {
+		return nil, fmt.Errorf("failed to load CA bundle for execProvider: %v", err)
+	}
+
+	var proxyURL string
+	if config.Proxy != nil {
+		url, err := config.Proxy(nil)
```

Comment:
shouldn't this be passing in the config host? also, Proxy can return a nil url with no error

```suggestion
hostURL, err := url.Parse(config.Host)
...
proxyURL, err := config.Proxy(hostURL)
...
if proxyURL != nil {
  ... = proxyURL.String()
}
```

Reply:
The `proxyURL` nil check makes sense, but I believe `config.Proxy()` accepts a `*http.Request`.

```golang
	Proxy func(*http.Request) (*url.URL, error)
```

Maybe `config.Proxy(&http.Request{URL: hostURL})`? This is kinda a tricky conversion, because we are essentially converting from a dynamic proxy URL (`rest.Config.Proxy`) to a static proxy URL (`clientauthentication.Client`). I would feel bad passing a half-baked `http.Request` to `config.Proxy`.

Reply:
mm, true. probably `http.NewRequest("", config.Host, nil)` to get the same normalization we'd get for actual requests.

Reply:
0750ed9c614